internal/datasource/config/config_writer.go

```diff
@@ -0,0 +1,190 @@
+				skippedFiles[fileData.Path] = struct{}{}
+				continue
+			}
+
+			fileDownloadResponse, err := cw.configClient.GetFile(fileData, filesUrl, tenantID.String())
+			if err != nil {
+				return fmt.Errorf("error getting file data from %s: %w", filesUrl, err)
+			}
+
+			err = writeFile(fileDownloadResponse.FileContent, fileDownloadResponse.FilePath)
+			if err != nil {
+				return fmt.Errorf("error writing to file %s: %w", fileDownloadResponse.FilePath, err)
+			}
+
+			currentFileCache[fileData.Path] = &instances.File{
+				Version:      fileData.Version,
+				Path:         fileData.Path,
+				LastModified: fileData.LastModified,
+			}
+		}
+	}
+
+	cw.currentFileCache = currentFileCache
+
+	return err
+}
+
+func (cw *ConfigWriter) Complete() error {
+	cache, err := json.MarshalIndent(cw.currentFileCache, "", "  ")
+	if err != nil {
+		return fmt.Errorf("error marshalling cache data from %s: %w", cw.cachePath, err)
+	}
+
+	err = writeFile(cache, cw.cachePath)
+	if err != nil {
+		return fmt.Errorf("error writing cache to %s: %w", cw.cachePath, err)
+	}
+
+	return err
+}
+
+func (cw *ConfigWriter) SetDataplaneConfig(dataplaneConfig nginx.DataplaneConfigInterface) error {
+	cw.dataplaneConfig = dataplaneConfig
+	return nil
+}
+
+func (cw *ConfigWriter) Reload() error {
+	return cw.dataplaneConfig.Reload()
+}
+
+func (cw *ConfigWriter) Validate() error {
+	return cw.dataplaneConfig.Validate()
+}
+
+func writeFile(fileContent []byte, filePath string) error {
+	if _, err := os.Stat(filePath); os.IsNotExist(err) {
+		slog.Debug("File does not exist, creating new file", "file", filePath)
+		err = os.MkdirAll(path.Dir(filePath), 0o750)
+		if err != nil {
+			return fmt.Errorf("error creating directory %s: %w", path.Dir(filePath), err)
+		}
+	}
+
+	err := os.WriteFile(filePath, fileContent, 0o644)
+	if err != nil {
+		return fmt.Errorf("error writing to file %s: %w", filePath, err)
+	}
+	slog.Debug("Content written to file", "filePath", filePath)
+
+	return nil
+}
+
+func readInstanceCache(cachePath string) (previousFileCache FileCache, err error) {
+	previousFileCache = FileCache{}
+
+	if _, err := os.Stat(cachePath); os.IsNotExist(err) {
+		return previousFileCache, fmt.Errorf("cache.json does not exist %s: %w", cachePath, err)
+	}
+
+	cacheData, err := os.ReadFile(cachePath)
+	if err != nil {
+		return previousFileCache, fmt.Errorf("error reading file cache.json %s: %w", cachePath, err)
+	}
+	err = json.Unmarshal(cacheData, &previousFileCache)
+	if err != nil {
+		return previousFileCache, fmt.Errorf("error unmarshalling data from cache.json %s: %w", cachePath, err)
+	}
+
+	return previousFileCache, err
+}
+
+func isFilePathValid(filePath string) (validPath bool) {
+	return filePath != "" && !strings.HasSuffix(filePath, "/")
+}
+
+func doesFileRequireUpdate(previousFileCache FileCache, fileData *instances.File) (updateRequired bool) {
+	if previousFileCache != nil && len(previousFileCache) > 0 {
```

Comment:
I think my IDE complained about this earlier, as a `nil` `map` is the same as `len(map) == 0`, so you could just simplify this to:
```suggestion
	if len(previousFileCache) > 0 {
```